File: SystemDesign/concepts/system-design/encryption/js/03_https_handshake.js
/**
 * HTTPS HANDSHAKE — Step by step simulation
 *
 * HTTPS = HTTP + TLS (Transport Layer Security)
 *
 * The handshake solves this problem:
 *   AES is fast but needs a shared key — how to share it safely?
 *   RSA can share keys safely but is too slow for bulk data.
 *   Solution: use RSA to share an AES key, then use AES for everything else.
 *
 * TLS 1.2 Handshake (RSA key exchange):
 *
 *   Client                              Server
 *     │                                   │
 *     │──── 1. ClientHello ──────────────▶│  "I support TLS 1.2, here are my ciphers"
 *     │                                   │
 *     │◀─── 2. ServerHello ───────────────│  "Use AES-256-GCM, here's my certificate"
 *     │◀─── 3. Certificate ───────────────│  Contains server's PUBLIC key
 *     │                                   │
 *     │  4. Verify certificate            │
 *     │     (is it signed by trusted CA?) │
 *     │                                   │
 *     │  5. Generate pre-master secret    │
 *     │──── 6. Encrypted pre-master ─────▶│  Encrypted with server's PUBLIC key
 *     │                                   │  Server decrypts with PRIVATE key
 *     │                                   │
 *     │  7. Both derive session key (AES) from pre-master secret
 *     │                                   │
 *     │──── 8. "Ready" (encrypted) ──────▶│
 *     │◀─── 9. "Ready" (encrypted) ───────│
 *     │                                   │
 *     │══════ All data encrypted with AES session key ══════│
 *
 * TLS 1.3 (modern — uses ECDHE instead of RSA for key exchange):
 *   Faster, Perfect Forward Secrecy — new key per session
 *   Even if private key is stolen later, past sessions can't be decrypted
 */

const crypto = require('crypto');

console.log('=== HTTPS TLS HANDSHAKE SIMULATION ===\n');

// ── Step 1 & 2: Hello messages ────────────────────────────────────────────────
console.log('STEP 1 — Client Hello');
console.log('  Client → Server: "I support TLS 1.2/1.3, AES-256-GCM, SHA-256"');
const clientRandom = crypto.randomBytes(28);
console.log('  Client random:', clientRandom.toString('hex').substring(0, 20) + '...');

console.log('\nSTEP 2 — Server Hello + Certificate');
console.log('  Server → Client: "Use AES-256-GCM. Here is my certificate."');

// ── Step 3: Server generates RSA key pair (represents certificate) ────────────
console.log('\nSTEP 3 — Server has RSA key pair (from certificate)');
const { publicKey: serverPublicKey, privateKey: serverPrivateKey } =
    crypto.generateKeyPairSync('rsa', {
        modulusLength: 2048,
        publicKeyEncoding:  { type: 'pkcs1', format: 'pem' },
        privateKeyEncoding: { type: 'pkcs1', format: 'pem' }
    });
console.log('  Server PUBLIC key sent to client (in certificate)');
console.log('  Server PRIVATE key stays on server — never leaves');

// ── Step 4: Client verifies certificate ───────────────────────────────────────
console.log('\nSTEP 4 — Client verifies certificate');
console.log('  Client checks: is this cert signed by a trusted CA?');
console.log('  (CA = Certificate Authority: DigiCert, Let\'s Encrypt, etc.)');
console.log('  Browser has built-in list of ~100 trusted CAs');
console.log('  If cert valid → continue. If not → "Your connection is not private" error');

// ── Step 5 & 6: Pre-master secret ─────────────────────────────────────────────
console.log('\nSTEP 5 — Client generates pre-master secret');
const preMasterSecret = crypto.randomBytes(46);  // 46 bytes
console.log('  Pre-master secret:', preMasterSecret.toString('hex').substring(0, 20) + '...');

console.log('\nSTEP 6 — Client encrypts pre-master with server\'s PUBLIC key');
const encryptedPreMaster = crypto.publicEncrypt(
    { key: serverPublicKey, padding: crypto.constants.RSA_PKCS1_OAEP_PADDING },
    preMasterSecret
);
console.log('  Encrypted pre-master sent to server');
console.log('  (Only server can decrypt — it has the private key)');

// ── Step 6b: Server decrypts ──────────────────────────────────────────────────
console.log('\nSTEP 6b — Server decrypts with PRIVATE key');
const decryptedPreMaster = crypto.privateDecrypt(
    { key: serverPrivateKey, padding: crypto.constants.RSA_PKCS1_OAEP_PADDING },
    encryptedPreMaster
);
console.log('  Match?', preMasterSecret.toString('hex') === decryptedPreMaster.toString('hex'));

// ── Step 7: Derive session key ────────────────────────────────────────────────
console.log('\nSTEP 7 — Both sides derive AES session key from pre-master secret');
// In real TLS: session key = PRF(pre_master_secret + client_random + server_random)
// Simplified here:
const sessionKey = crypto.createHash('sha256')
    .update(Buffer.concat([preMasterSecret, clientRandom]))
    .digest();
console.log('  Session key (AES-256):', sessionKey.toString('hex').substring(0, 20) + '...');
console.log('  Both client and server now have the SAME AES key');
console.log('  No attacker can derive it — pre-master was RSA-encrypted');

// ── Step 8 & 9: Ready ────────────────────────────────────────────────────────
console.log('\nSTEP 8-9 — Finished messages (verify handshake integrity)');
console.log('  Both send "Finished" encrypted with session key');
console.log('  Confirms: both sides derived same key correctly');

// ── Step 10: Data transfer with AES ──────────────────────────────────────────
console.log('\nSTEP 10 — Data transfer using AES session key');

function aesEncrypt(text, key) {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);
    let enc = cipher.update(text, 'utf8', 'hex');
    enc += cipher.final('hex');
    return { iv: iv.toString('hex'), ciphertext: enc, tag: cipher.getAuthTag().toString('hex') };
}

function aesDecrypt(enc, key) {
    const decipher = crypto.createDecipheriv('aes-256-gcm', key, Buffer.from(enc.iv, 'hex'));
    decipher.setAuthTag(Buffer.from(enc.tag, 'hex'));
    let dec = decipher.update(enc.ciphertext, 'hex', 'utf8');
    dec += decipher.final('utf8');
    return dec;
}

const httpRequest  = 'GET /api/account HTTP/1.1\nAuthorization: Bearer secret-token-123';
const httpResponse = '{"balance": 50000, "account": "savings"}';

const encReq  = aesEncrypt(httpRequest, sessionKey);
const encResp = aesEncrypt(httpResponse, sessionKey);

console.log('  Client sends (encrypted):', encReq.ciphertext.substring(0, 40) + '...');
console.log('  Server decrypts request: ', aesDecrypt(encReq, sessionKey).substring(0, 40));
console.log('  Server sends (encrypted):', encResp.ciphertext.substring(0, 40) + '...');
console.log('  Client decrypts response:', aesDecrypt(encResp, sessionKey));

console.log('\n=== SUMMARY ===');
console.log('RSA used for:  key exchange only (1 time, during handshake)');
console.log('AES used for:  all actual data (fast, entire session)');
console.log('Why hybrid?    RSA = secure but slow | AES = fast but needs shared key');
console.log('               RSA solves AES key sharing problem');

File: SystemDesign/concepts/system-design/encryption/python/03_https_handshake.py
"""
HTTPS HANDSHAKE SIMULATION — TLS 1.2 step by step

WHAT HAPPENS when browser opens https://google.com?
  1. Client says hello (what ciphers I support)
  2. Server sends its certificate (contains public key)
  3. Client verifies certificate (signed by trusted CA?)
  4. Client generates a secret, encrypts with server's PUBLIC key
  5. Server decrypts with PRIVATE key → both have same secret
  6. Both derive an AES session key from that secret
  7. All HTTP data is now AES-encrypted

WHY HYBRID (RSA + AES)?
  RSA: secure key exchange, but ~1000x slower than AES → can't encrypt all data
  AES: blazing fast, but needs a shared key
  Together: RSA safely shares the AES key, AES encrypts everything else

Install: pip install cryptography
"""

import os
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

print("=" * 60)
print("TLS 1.2 Handshake Simulation")
print("=" * 60)

# ── STEP 1: Server has a certificate (RSA key pair) ───────────────────────────
print("\n[SERVER] Generating RSA-2048 key pair (normally done once, stored in cert)...")
server_private = rsa.generate_private_key(public_exponent=65537, key_size=2048)
server_public  = server_private.public_key()
print("[SERVER] Ready. Public key will be sent in certificate.")

# ── STEP 2: ClientHello ────────────────────────────────────────────────────────
print("\n[CLIENT → SERVER] Step 1: ClientHello")
client_random = os.urandom(28)     # 28 random bytes + 4 bytes timestamp
print(f"  client_random: {client_random.hex()[:20]}...")
print("  Supported ciphers: TLS_RSA_WITH_AES_256_GCM_SHA384")

# ── STEP 3: ServerHello + Certificate ─────────────────────────────────────────
print("\n[SERVER → CLIENT] Step 2: ServerHello + Certificate")
server_random = os.urandom(32)
print(f"  server_random: {server_random.hex()[:20]}...")
print("  Chosen cipher: TLS_RSA_WITH_AES_256_GCM_SHA384")
print("  Sending certificate with PUBLIC KEY...")
# In real TLS: certificate is signed by a CA (Certificate Authority)
# Browser has CA's public key pre-installed → can verify the cert

# ── STEP 4: Client verifies certificate ───────────────────────────────────────
print("\n[CLIENT] Step 3: Verify certificate")
print("  Checking: Is this cert signed by a trusted CA? (simulated: YES)")
print("  Checking: Is hostname matching? (simulated: YES)")
print("  Certificate trusted. Proceeding.")

# ── STEP 5: Client generates pre-master secret, encrypts with server's pub key
print("\n[CLIENT] Step 4: Generate pre-master secret")
pre_master_secret = os.urandom(48)   # 48 random bytes
print(f"  pre_master_secret: {pre_master_secret.hex()[:20]}... (NEVER sent in plaintext)")

print("\n[CLIENT → SERVER] Step 5: Send encrypted pre-master secret")
encrypted_pms = server_public.encrypt(
    pre_master_secret,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
print(f"  encrypted_pms: {encrypted_pms.hex()[:30]}... (RSA-encrypted, safe to send)")

# ── STEP 6: Server decrypts pre-master secret ─────────────────────────────────
print("\n[SERVER] Step 6: Decrypt pre-master secret with PRIVATE key")
decrypted_pms = server_private.decrypt(
    encrypted_pms,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
print(f"  Decrypted pre_master_secret: {decrypted_pms.hex()[:20]}...")
print(f"  Match? {decrypted_pms == pre_master_secret}")

# ── STEP 7: Both sides derive the same AES session key ────────────────────────
print("\n[CLIENT + SERVER] Step 7: Derive AES session key (same on both sides)")
# Both have: pre_master_secret + client_random + server_random
# PRF (Pseudo-Random Function) combines them into a master secret
master_secret_input = pre_master_secret + client_random + server_random

# Use HKDF to derive a 32-byte AES key (simulates TLS PRF)
hkdf = HKDF(algorithm=hashes.SHA256(), length=32, salt=None, info=b"tls session key")
session_key = hkdf.derive(master_secret_input)

print(f"  session_key: {session_key.hex()[:30]}...")
print("  CLIENT has this key. SERVER has this key. Attacker does NOT.")

# ── STEP 8: Encrypted communication ───────────────────────────────────────────
print("\n[CLIENT → SERVER] Step 8: HTTP request (AES-256-GCM encrypted)")
aesgcm = AESGCM(session_key)
nonce = os.urandom(12)

http_request = b"GET /account HTTP/1.1\r\nHost: bank.com\r\nCookie: session=abc123"
encrypted_request = aesgcm.encrypt(nonce, http_request, None)
print(f"  Encrypted: {encrypted_request.hex()[:40]}...")

# Server decrypts
http_received = aesgcm.decrypt(nonce, encrypted_request, None)
print(f"\n[SERVER] Decrypted request: {http_received.decode()[:50]}...")

print("\n" + "=" * 60)
print("HANDSHAKE COMPLETE")
print("=" * 60)
print("""
Summary:
  RSA (asymmetric): used ONCE to exchange the pre-master secret
  AES (symmetric):  used for ALL data (fast, stream encryption)

  Even if attacker records the encrypted traffic, they need:
  → server's private key to get pre_master_secret
  → which lets them derive session_key
  → which lets them decrypt everything

  TLS 1.3 improvement: uses ECDHE (Elliptic Curve Diffie-Hellman)
  → new key pair per session → forward secrecy
  → even if private key is stolen later, past sessions stay safe
  → see 04_rsa_vs_ecc.py
""")
